features/cellular/framework/mux/cellular_mux.cpp

```diff
@@ -0,0 +1,1253 @@
+#define DM_FRAME_LEN                        6u        /* Length of the DM frame in number of bytes. */
+#define UIH_FRAME_MIN_LEN                   6u        /* Minimum length of user frame. */
+
+#define T1_TIMER_VALUE                      300u      /* T1 timer value. */
+#define RETRANSMIT_COUNT                    3u        /* Retransmission count for the tx frames requiring a response. */
+
+#define FRAME_TYPE_SABM                     0x2Fu     /* SABM frame type coding in the frame control field. */
+#define FRAME_TYPE_UA                       0x63u     /* UA frame type coding in the frame control field. */
+#define FRAME_TYPE_DM                       0x0Fu     /* DM frame type coding in the frame control field. */
+#define FRAME_TYPE_DISC                     0x43u     /* DISC frame type coding in the frame control field. */
+#define FRAME_TYPE_UIH                      0xEFu     /* UIH frame type coding in the frame control field. */
+
+#define PF_BIT                              (1u << 4) /* P/F bit position in the frame control field. */
+#define EA_BIT                              (1u << 0) /* E/A bit position in the frame address field. */
+#define CR_BIT                              (1u << 1) /* C/R bit position in the frame address field. */
+
+#define DLCI_ID_LOWER_BOUND                 1u        /* Lower bound value of DLCI id.*/
+#define DLCI_ID_UPPER_BOUND                 63u       /* Upper bound value of DLCI id.*/
+
+#define FRAME_FLAG_SEQUENCE_FIELD_INDEX     0         /* Index of the frame flag sequence field. */
+#define FRAME_ADDRESS_FIELD_INDEX           1u        /* Index of the frame address field. */
+#define FRAME_CONTROL_FIELD_INDEX           2u        /* Index of the frame control field. */
+#define FRAME_LENGTH_FIELD_INDEX            3u        /* Index of the frame length indicator field. */
+#define FRAME_INFORMATION_FIELD_INDEX       4u        /* Index of the frame information field. */
+
+#define FRAME_START_READ_LEN                1u        /* Frame start read length in number of bytes. */
+#define FRAME_HEADER_READ_LEN               3u        /* Frame header read length in number of bytes. */
+#define FRAME_TRAILER_LEN                   2u        /* Frame trailer read length in number of bytes. */
+
+#define LENGTH_INDICATOR_OCTET              1u        /* Length indicator field value used in frame. */
+
+/* Definition for frame header type. */
+struct frame_hdr_t {
+    uint8_t flag_seq;       /* Flag sequence field. */
+    uint8_t address;        /* Address field. */
+    uint8_t control;        /* Control field. */
+    uint8_t length;         /* Length field. */
+    uint8_t information[1]; /* Begin of the information field if present. */
+};
+
+uint8_t Mux::_shared_memory = 0;
+FileHandle* Mux::_serial    = NULL;
+EventQueue* Mux::_event_q   = NULL;
+
+MuxDataService Mux::_mux_objects[MBED_CONF_MUX_DLCI_COUNT];
+Semaphore Mux::_semaphore(0);
+PlatformMutex Mux::_mutex;
```

Comment:
I've noticed few global variables in some files, are all used always ? I am asking because linker might not remove these even if not used (we provide `SingletonPtr` that helps in this cases)

Reply:
The linker will definitely not remove these - it will link them in so it can run their constructors.

It would be nice to have a CI test to warn about this. Give a warning like "this PR makes example-blinky bigger - did you mean to do that?"

In the absence of that, and given that static singletons are a common pattern in this PR, it's probably a good idea for the developers to actually run that test manually - build blinky example with this PR in and out and make sure the ROM and RAM isn't getting any bigger to flush out any other instances of this.

The fundamental problem is that the entire Mux class is static, for no obvious reason. I'd rather fix that than scatter it with `SingletonPtr`, digging deeper.

I can't see why it's not just a normal object. The LoRaWAN stuff had the same design error, I recall - that got largely fixed apart from an internal issue with an underlying singleton C library. I believe the final state was "looks like a normal C++ object, so you construct it like one; constructing 2 won't actually work yet due to the internal limitation, but it could in future". I see no external dependencies preventing that this time, so any reason you can't just strip all the statics out here, and just construct a `Mux` object? 

That will eliminate this static data problem as long as the application doesn't have a `Mux`.

If the expectation is you'll only ever have 1 mux, then it doesn't make any difference whether the members are static or not - you may as well just make the thing a "normal" C object. The static should only kick in for optimisations like "share the crc table", and even that would only matter if someone was using 2 muxes.

Side note - I got called out for making an entirely static "object" -  https://github.com/ARMmbed/mbed-os/pull/5419#pullrequestreview-74057050 . It  was the case there that the `Kernel` is not really an object, so it ended up as a namespace - matching the style of `this_thread` in C++11 for example. In your case though a `Mux` would very naturally be an object - there's no reason to block multiple instantiation.

Reply:
Discussed with Antti - I'd be okay as a "quick fix" making the Mux files conditionally compiled via a JSON setting. They're not currently required for the basic platforms we're testing against, as I understand it. Would that do for you, @0xc0170 ?

Reply:
Anwers to Qs made above:
Q:  I see no external dependencies preventing that this time, so any reason you can't just strip all the statics out here, and just construct a Mux object?
A: Should not be any reason against this approach. Task IOTCELL-616 created for tracking.

Qs for feedback originator:
Statemant was made: "The LoRaWAN stuff had the same design error "
Q: Could you provide the link to the material, which collects these design guidelines for the development work so we can avoid these errors in the future.

Reply:
Design guidelines are here, but I don't see a specific reference calling out the static trap.

https://os.mbed.com/docs/v5.7/reference/guidelines.html

Closest I can see is "Each module should provide an object-oriented C++ user API." Which a static design isn't. And it's only a "should"

I think it would be worth putting an explicit statement about why static C++ objects must be avoided and pointing at SingletonPtr. (Although I don't think SingletonPtr will actually work for static members) 

It would be even more useful to have the CI spotting it, as not everyone is going to be going through guideline docs with a fine-tooth comb, or even necessarily finding them.

Do you reckon you can strip the statics in time for merge, or do we still need a plan B?

BTW, I believe should be no problem with the CRC table being const static, as it's plain data. It's only static C++ objects with real constructors that have to be avoided.

Reply:
Regarding Q:
"Do you reckon you can strip the statics in time for merge, or do we still need a plan B?"
It was decided by Antti that for now we will go with the conditional compilation plan.

Reply:
Conditional compilation is fine by me. On naming, I suggest `cellular.mux-enabled`, to match `lwip.ipv6-enabled` or `lwip.ppp-enabled`.

The only problem is then that making it non-static would be a visible API change in a future release.

There's very strong reluctance to actually remove or even deprecate APIs people might be relying on once they land in master - (see https://github.com/ARMmbed/mbed-os/pull/5420#pullrequestreview-80086519).

You might be able to pre-empt any complaints about reworking it for 5.9 if the current `Mux` API was somehow flagged as "internal", "draft" or "subject to change" or something. @0xc0170 - does that make sense?

Reply:
Added mux-enabled flag in cellular mbed_lib.json, which by default out-comments MUX 
https://github.com/AriParkkila/mbed-os/commit/78dd5117c8a4438d5fa7eb5d7733ad6f7598440d